src/app/shared/services/command.service.ts

```diff
@@ -0,0 +1,295 @@
+      clearInterval(redeem?.timer);
+      this.autoredeems.delete(autoID);
+    }
+  }
+
+  handleCommand(command: Commands, permissions: UserPermissions, text: string, chainCommand = false) {
+    const hasPerms = this.chatService.hasChatCommandPermissions({ permissions }, command.permissions);
+
+    if (!hasPerms || this.cooldowns.get(command.id)?.onCooldown) {
+      return;
+    }
+
+    switch (command.type) {
+      case 'counter':
+        this.handleCounterCommand(command);
+        break;
+      case 'sound':
+        this.handleSoundCommand(command);
+        break;
+      case 'chat':
+        this.handleChatCommand(command, text);
+        break;
+      case 'choice':
+        this.handleChoiceCommand(command);
+        break;
+      default:
+        return this.logService.add(
+          `Tried handling an unsupported command type.\n${JSON.stringify(command, null, 2)}`,
+          'error',
+          'CommandService.handleCommand',
+        );
+    }
+
+    /**
+     * If we get here, we can assume we ran a correct command type.
+     * Let's handle cooldowns if any exist.
+     */
+    const duration = command.cooldown * 1000;
+
+    this.cooldowns.set(command.id, {
+      duration,
+      onCooldown: true,
+      timeout: setTimeout(() => this.cooldowns.set(command.id, { duration, onCooldown: false }), duration),
+    });
+
+    this.logService.add(`Processing command.\n${JSON.stringify(command, null, 2)}`, 'info', 'CommandService.handleCommand');
+
+    /**
+     * If the command isn't a chain command call then it's safe to assume we want to check the commands
+     * possible chain commands.
+     */
+    if (!chainCommand) {
+      this.handleCommandChain(command, permissions, text);
+    }
+  }
+
+  handleCommandChain(command: Commands, permissions: UserPermissions, text: string) {
+    const { chainCommands } = command;
+
+    // Obviously, if there are no chain commands we want to ignore this.
+    if (!chainCommands.length) {
+      return;
+    }
+
+    this.commands$
+      .pipe(
+        first(),
+        takeUntilDestroyed(this.destroyRef),
+        map(commands => chainCommands.map(
+          chain => ({
+            command: commands.find(c => c.id === chain.chainCommandID),
+            chainCommandDelay: chain.chainCommandDelay,
+          }),
+        )),
+        map(commands => commands.filter(c => !!c.command)),
+      )
+      .subscribe(commands => {
+        /**
+         * I probably should handle deleting the ID if the command doesn't exist.
+         */
+        if (!commands.length) {
+          return;
+        }
+
+        let durationSum = 0;
+
+        for (const chainCommand of commands) {
+          const { command, chainCommandDelay } = chainCommand;
+
+          if (!command) {
+            continue;
+          }
+
+          /**
+           * This is a monstrous creation that shouldn't exist
+           * I'm way to lazy to await with promises here
```

Comment:
Move this stuff into a switchMap above, change your forLoop to construct an array of promises, then combineLatest the promise array. If you do this, just make sure to check for an empty array!